docs/rfc/004-integer-type-casts.md

```diff
@@ -0,0 +1,231 @@
+- Alessandro Coglio
+- Eric McCarthy
+- Jon Pavlik
+- Damir Shamanaev
+- Damon Sicore
+- Howard Wu
+
+## Status
+
+DRAFT
+
+# Summary
+
+This proposal provides support for casts among integer types in Leo.
+The syntax is similar to Rust.
+Two possible semantics are discussed:
+_value-preserving casts_,
+which just serve to change types
+but cause errors when values are not representable in the new types;
+and _values-changing casts_,
+which never cause errors but may change the mathematical values.
+
+# Motivation
+
+Currently the Leo integer types are "siloed":
+arithmetic integer operations require operands of the same type
+and return results of the same type.
+There are no implicit or explicit ways to turn, for example,
+a `u8` into a `u16`, even though
+every non-negative integers that fits in 8 bits also fits in 16 bits.
+However, the ability to convert values between different (integer) types
+is a useful feature that is normally found in programming languages.
+
+# Design
+
+## Background
+
+Recall that Leo supports the following _integer types_:
+```
+u8 u16 u32 u64 u128
+i8 i16 i32 i64 i128
+```
+
+## Scope
+
+This RFC proposes type casts between any two integer types,
+but not between two non-integer types
+or between an integer type and a non-integer type.
+
+This RFC does not propose any implicit cast,
+even widening casts (i.e. upcasts)
+from a type to another type with the same signedness
+and with the same or larger size
+(e.g. from `u8` to `u16`).
+All the type casts must be explicit.
+
+## Syntax and Static Semantics
+
+The proposed syntax is
+```
+<expression> as <integer-type>
+```
+where `<expression>` must have an integer type.
+
+The ABNF grammar is modified as follows:
+```
+; add this rule:
+cast-expression = unary-expression
+                / cast-expression %s"as" integer-type
+
+; modify this rule:
+exponential-expression = cast-expression
+                       / cast-expression "**" exponential-expression
+```
+There is no need to modify the `keyword` rule
+because it already includes `as` as one of the keywords.
+Note the use of `integer-type` in the `cast-expression` rule;
+an alternative is to use `type` there
+and check post-parsing that the type is in fact an integer one.
+
+The above grammar rules imply that casts bind
+tighter than binary operators and looser than unary operators.
+For instance,
+```
+x + - y as u8
+```
+is like
+```
+x + ((- y) as u8)
+```
+This precedence is the same as in Rust:
+see [here](https://doc.rust-lang.org/stable/reference/expressions.html#expression-precedence).
+
+## Dynamic Semantics
+
+When the mathematical integer value of the expression
+is representable in the type that the expression is cast to,
+there is no question that the cast must succeed
+and merely change the type of the Leo value,
+but not its mathematical integer value.
+This is always the case when the cast is to a type
+with the same signedness and with the same or larger size.
+This is sometimes the case when
```

Comment:
A little unclear.  Isn't it _always_ the case that the `mathematical integer value of the expression
is representable in the type that the expression is cast to` when `the value in question is in the intersection of the two ranges`?  I probably would write `This is also the case...` instead of `This is sometimes the case...`

Reply:
I agree, I changed it as you suggest.